tests/test_mains/test_dedupe.py

```diff
@@ -0,0 +1,166 @@
+#!/usr/bin/env python3
+# encoding: utf-8
+
+from nose import with_setup
+from nose.tools import make_decorator
+from nose.plugins.skip import SkipTest
+from contextlib import contextmanager
+import psutil
+
+from tests.utils import *
+
+REFLINK_CAPABLE_FILESYSTEMS = {'btrfs', 'xfs', 'ocfs2'}
+
+@contextmanager
+def assert_exit_code(status_code):
+    """
+    Assert that the with block yields a subprocess.CalledProcessError
+    with a certain return code. If nothing is thrown, status_code
+    is required to be 0 to survive the test.
+    """
+    try:
+        yield
+    except subprocess.CalledProcessError as exc:
+        assert exc.returncode == status_code
+    else:
+        # No exception? status_code should be fine.
+        assert status_code == 0
+
+
+def is_on_reflink_fs(path):
+    parts = psutil.disk_partitions(all=True)
+
+    # iterate up from `path` until mountpoint found
+    p = path
+    while 1:
+        match = next((x for x in parts if x.mountpoint == p), None)
```

Comment:
I like your cleverness here, but Python is made to be , therefore I think something like this might be read easier:

```python
def up(path):
    while path:
        yield path
        path = os.path.dirname(path)
        if path == "/":
            break

def is_on_reflink_fs(path):
    for part in psutil.disk_partitions(all=True):
        for up_path in up(path):
            if part.mountpoint == up_path:
                print("{0} is {1} mounted at {2}".format(path, match.fstype, p))
                return match.fstype in REFLINK_CAPABLE_FILESYSTEMS
    print("no mountpoint found for {0}".format(path))
    return False
````
(Not tested)

Also we might want to consider moving things like ``needs_reflink_fs`` and ``assert_exit_code`` to utils.